### PR TITLE
Feature/parquet append

### PR DESCRIPTION
**Pull request by Talmaj**
Added append functionality for dask.DataFrame.to_parquet #1936. @martindurant
- columns equality check of existing parquet file and appended df
- divisions overlap check of existing parquet file and appended df
- basic test of append functionality
- documentation

### CODE REVIEW

File: dask/dataframe/io/parquet.py [hunk 1]
```diff
@@ -229,8 +239,33 @@ def to_parquet(path, df, compression=None, write_index=None, has_nulls=None,
...
+    if append:
+        pf = fastparquet.api.ParquetFile(path, open_with=myopen)
+        if pf.file_scheme != 'hive':
+            raise ValueError('Requested file scheme is hive, '
+                             'but existing file scheme is not.')
+
+        assert (all(pf.columns == df._meta.columns),
+                'Appended columns not exactly the same.\n'
+                'New: {} | Previous: {}'.
+                format(pf.columns,list(df._meta.columns)))
```
Comment (Talmaj) [inline dask/dataframe/io/parquet.py]: If we have `write_index == True`, then we do `df = df.reset_index()` and `df._meta.columns` contains the index as well. If we don't write the index then the parquet file shouldn't contain one as well. I assumed we are appending dfs with the same properties.

Comment (mrocklin) [inline dask/dataframe/io/parquet.py]: Ah, great then.  As always, I would be more comfortable if we had tests for both indexed an non-indexed behavior.

Comment (Talmaj) [inline dask/dataframe/io/parquet.py]: I will add more tests.

**Comment by martindurant** [comment]
Thanks for looking into this, @Talmaj; I may not be able to review your work for a couple of days.

File: dask/dataframe/io/parquet.py [hunk 2]
```diff
@@ -229,8 +239,33 @@ def to_parquet(path, df, compression=None, write_index=None, has_nulls=None,
...
+        minmax = fastparquet.api.sorted_partitioned_columns(pf)
+        divisions = list(minmax[index_col]['min']) + [
+            minmax[index_col]['max'][-1]]
+        fmd = pf.fmd
+        i_offset = fastparquet.writer.find_max_part(fmd.row_groups)
+        if new_divisions[0] < divisions[-1] and not ignore_divisions:
+            raise ValueError(
+                'Appended divisions overlapping with the previous ones.\n'
+                'New: {} | Previous: {}'
+                .format(divisions[-1], new_divisions[0]))
```
Comment (Talmaj) [inline dask/dataframe/io/parquet.py]: I added more tests including for exceptions.

**Comment by mrocklin** [comment]
@Talmaj I'd prefer to wait until @martindurant has had a chance to look things over.  Are you OK waiting on this for a bit?

**Comment by Talmaj** [comment]
@mrocklin sure, I'll wait.

File: dask/dataframe/io/parquet.py [hunk 3]
```diff
@@ -196,6 +197,13 @@ def to_parquet(path, df, compression=None, write_index=None, has_nulls=None,
         encoding is applied to all object columns.
     storage_options : dict
         Key/value pairs to be passed on to the file-system backend, if any.
+    append: bool (False)
+        If False, construct data-set from scratch; if True, add new row-group(s)
+        to existing data-set. In the latter case, the data-set must exist,
+        and the schema must match the input data.
+    ignore_divisions: bool (False)
+        If False raises error when previous divisions overlap with the new
+        appended divisions.
```
Comment (martindurant) [inline dask/dataframe/io/parquet.py]: (ignored if append=False)

File: dask/dataframe/io/parquet.py [hunk 4]
```diff
@@ -229,8 +241,38 @@ def to_parquet(path, df, compression=None, write_index=None, has_nulls=None,
                                            fixed_text=fixed_text,
                                            object_encoding=object_encoding)
 
+    if append:
+        pf = fastparquet.api.ParquetFile(path, open_with=myopen)
+        if pf.file_scheme != 'hive':
+            raise ValueError('Requested file scheme is hive, '
+                             'but existing file scheme is not.')
+
+        if len(set(pf.columns) ^ set(df._meta.columns)):
```
Comment (martindurant) [inline dask/dataframe/io/parquet.py]: not the same as `set(pf.columns) != set(df._meta.columns)`

File: dask/dataframe/io/parquet.py [hunk 5]
```diff
@@ -229,8 +241,38 @@ def to_parquet(path, df, compression=None, write_index=None, has_nulls=None,
...
+            raise ValueError('Appended columns not the same.\n'
+                             'New: {} | Previous: {}'
+                             .format(pf.columns, list(df._meta.columns)))
+        else:
+            df = df[pf.columns]
+
+        fmd = pf.fmd
```
Comment (martindurant) [inline dask/dataframe/io/parquet.py]: Here we blew away the `fmd` generated from the data itself. Checks around the requested dtypes should go here: an exception should be raised not only if column names mis-match, but also if the inferred schemas do.

File: dask/dataframe/io/tests/test_parquet.py [hunk 6]
```diff
@@ -136,6 +136,90 @@ def test_categorical():
...
+        ddf2 = dd.from_pandas(df2, chunksize=2)
+
+        ddf1.to_parquet(tmp)
+
+        with pytest.raises(ValueError) as excinfo:
+            ddf2.to_parquet(tmp, append=True)
+
+        assert 'Appended columns' in str(excinfo.value)
+
+
```
Comment (martindurant) [inline dask/dataframe/io/tests/test_parquet.py]: + test where we have the same column names, but not the same dtypes, should raise.

**Comment by martindurant** [comment]
I have had a look through, and I am generally approving of the code. Just a few comments above. The only further comment, there is some repetition of the code in fastparquet's own append, and I wonder to what extent that is avoidable. I am not troubled to try to factor out the common bits right now.

File: dask/dataframe/io/parquet.py (hunk 5 above)
Comment (Talmaj) [inline dask/dataframe/io/parquet.py]: Do I need to check complete schema (`fmd.schema != pf.fmd.schema`) or is it enough to only check the column names and dtypes? Namely I left the option that columns don't need to be sorted correctly, and reorder them if their naming and dtypes are correct.

**Comment by martindurant** [comment]
I am unsure whether there is any requirement for the schema to be identical, but I'm happy to go with the dtypes check for now. It could plausibly be an issue eventually, but I don't immediately see how.

+1

**Comment by mrocklin** [comment]
Thank you for taking this on and seeing it through @Talmaj .  I'm very glad to see this in.

**Comment by Talmaj** [comment]
You're welcome. Glad to help and contribute.

**Pull request merged**